src/Features/Core/Portable/CodeRefactorings/AddMissingImports/AbstractAddMissingImportsFeatureService.cs

```diff
@@ -0,0 +1,150 @@
+            var diagnostics = await GetDiagnosticsAsync(document, textSpan, cancellationToken).ConfigureAwait(false);
+            if (diagnostics.IsEmpty)
+            {
+                return document.Project;
+            }
+
+            // Find fixes for the diagnostic where there is only a single fix.
+            var usableFixes = await GetUnambiguousFixesAsync(document, diagnostics, cancellationToken).ConfigureAwait(false);
+            if (usableFixes.IsEmpty)
+            {
+                return document.Project;
+            }
+
+            // Apply those fixes to the document.
+            var newDocument = await ApplyFixesAsync(document, usableFixes, cancellationToken).ConfigureAwait(false);
+            return newDocument.Project;
+        }
+
+        private async Task<ImmutableArray<Diagnostic>> GetDiagnosticsAsync(Document document, TextSpan textSpan, CancellationToken cancellationToken)
+        {
+            var project = document.Project;
+            var compilation = await project.GetCompilationAsync(cancellationToken);
+
+            // Ensure that diagnostics for this document are always in document location
+            // order.  This provides a consistent and deterministic order for fixers
+            // that want to update a document.
+            var diagnostics = compilation.GetDiagnostics(cancellationToken)
+                .Where(diagnostic => FixableDiagnosticIds.Contains(diagnostic.Id)
+                    && diagnostic.Location.SourceTree.FilePath == document.FilePath
+                    && diagnostic.Location.SourceSpan.IntersectsWith(textSpan))
+                .ToImmutableArray()
+                .Sort((d1, d2) => d1.Location.SourceSpan.Start - d2.Location.SourceSpan.Start);
+
+            return diagnostics;
+        }
+
+        private async Task<ImmutableArray<AddImportFixData>> GetUnambiguousFixesAsync(Document document, ImmutableArray<Diagnostic> diagnostics, CancellationToken cancellationToken)
+        {
+            var solution = document.Project.Solution;
+            var symbolSearchService = solution.Workspace.Services.GetService<ISymbolSearchService>();
+            // Since we are not currently considering NuGet packages, pass an empty array
+            var packageSources = ImmutableArray<PackageSource>.Empty;
+            var addImportService = document.GetLanguageService<IAddImportFeatureService>();
+
+            var getFixesForDiagnosticsTasks = diagnostics
+                .GroupBy(diagnostic => diagnostic.Location.SourceSpan)
+                .Select(diagnosticsForSourceSpan => addImportService
+                    .GetFixesForDiagnosticsAsync(document, diagnosticsForSourceSpan.Key, diagnosticsForSourceSpan.AsImmutable(),
+                        symbolSearchService, searchReferenceAssemblies: true, packageSources, cancellationToken));
+
+            await Task.WhenAll(getFixesForDiagnosticsTasks).ConfigureAwait(false);
```

Comment:
@sharwell pointed out that teh when-all is not necessary.  these tasks are naturally running in parallel, and it's fine to just have the foreach loop below that does hte awaiting of individual tasks.  worst case is that one loop waits a while for that task to finish. but that's what WhenAll does anyways :)